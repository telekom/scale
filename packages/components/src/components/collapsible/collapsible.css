/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --spacing-y: var(--telekom-spacing-unit-x4);

  /* button */
  --font-size-button: var(--telekom-typography-font-size-callout);
  --font-weight-button: var(--telekom-typography-font-weight-extra-bold);
  --line-height-button: var(--telekom-typography-line-spacing-standard);
  --border-width-button: var(--telekom-spacing-unit-x05);
  --radius-button: var(--telekom-radius-small);
  --spacing-left-button: var(--telekom-spacing-unit-x1);
  --spacing-left-button-text: var(--telekom-spacing-unit-x2);
  --color-button-hover: var(--telekom-color-text-and-icon-primary-hovered);
  --color-button-active: var(--telekom-color-text-and-icon-primary-pressed);
  --background-button-hover: var(--telekom-color-background-canvas-subtle);
  --border-color-button-focus: var(--telekom-color-functional-focus);

  /* icon */
  --spacing-top-icon: calc(var(--telekom-spacing-unit-x1) + 1px);

  /* content */
  --spacing-right-content: var(--telekom-spacing-unit-x12);
  --spacing-left-content: var(--telekom-spacing-unit-x8);
  --spacing-top-content: var(--telekom-spacing-unit-x4);
  --spacing-bottom-content: var(--telekom-spacing-unit-x6);
}

.collapsible {
  position: relative;
  margin: var(--spacing-y) 0;
}

.collapsible--expanded:before {
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  border: var(--telekom-spacing-unit-x025) solid transparent;
  content: '';
  display: block;
  position: absolute;
  pointer-events: none;
  bottom: calc(-1 * var(--spacing-y));
}

.collapsible__heading {
  margin: 0;
}

.collapsible__button {
  color: currentColor;
  width: 100%;
  display: flex;
  text-align: left;
  align-items: flex-start;
  padding-top: 0;
  border-color: transparent;
  border-style: solid;
  padding-right: 0;
  padding-bottom: 0;
  background-color: transparent;
  font-size: var(--font-size-button);
  font-weight: var(--font-weight-button);
  line-height: var(--line-height-button);
  border-width: var(--border-width-button);
  border-radius: var(--radius-button);
  padding-left: var(--spacing-left-button);
  /* a11y: allows some "external" typographic styles to leak in
   https://www.w3.org/TR/WCAG21/#text-spacing */
  font-family: inherit;
  word-spacing: inherit;
  letter-spacing: inherit;
}

.collapsible__button-text {
  margin-left: var(--spacing-left-button-text);
}

.collapsible__button:hover {
  cursor: pointer;
  color: var(--color-button-hover);
  background-color: var(--background-button-hover);
}

.collapsible__button:active {
  color: var(--color-button-active);
}

.collapsible__button:focus {
  outline: none;
  border-color: var(--border-color-button-focus);
}

.collapsible__icon {
  transform: rotate(0.75turn);
  flex-shrink: 0;
  margin-top: var(--spacing-top-icon);
}

.collapsible__button[aria-expanded='true'] .collapsible__icon {
  transform: none;
}

.collapsible__content {
  padding-right: var(--spacing-right-content);
  padding-left: var(--spacing-left-content);
  /* These should collapse with tags like p, that's what we want */
  margin-top: var(--spacing-top-content);
  margin-bottom: var(--spacing-bottom-content);
}
