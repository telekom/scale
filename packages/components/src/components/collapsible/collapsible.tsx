/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  h,
  Prop,
  Host,
  Event,
  State,
  Element,
  EventEmitter,
} from '@stencil/core';
import classNames from 'classnames';

export interface CollapsibleEventDetail {
  expanded: boolean;
}

let i = 0;

@Component({
  tag: 'scale-collapsible',
  styleUrl: './collapsible.css',
  shadow: true,
})
export class Collapsible {
  headingElement: HTMLElement;
  headingId: string;
  panelId: string;

  @Element() el: HTMLElement;

  /** Set to `true` to expand */
  @Prop({ mutable: true, reflect: true }) expanded: boolean;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** Default aria-level for heading */
  @State() level: number = 2;

  /** Emitted so parent <scale-accordion> knows about it */
  @Event() scaleExpand: EventEmitter<CollapsibleEventDetail>;

  componentWillLoad() {
    const j = i++;
    this.headingId = 'collapsable-heading-' + j;
    this.panelId = 'collapsable-panel-' + j;
  }

  componentDidLoad() {
    this.setHeadingFromLightDOM();
  }

  /**
   * In this method we:
   * - query the first element from the light DOM, it should be a heading (e.g. h2)
   * - take its content and place it into our own heading element
   * - set aria-level to the level of that provided in the light DOM
   * - remove the original heading
   * @see https://inclusive-components.design/collapsible-sections/
   */
  setHeadingFromLightDOM() {
    const lightHeading = this.el.querySelector(':first-child');
    if (lightHeading == null) {
      return;
    }
    const level = parseInt(lightHeading.tagName.substr(1), 10);

    if (!level) {
      // tslint:disable-next-line
      console.warn(
        'The first element inside each <scale-collapsible> should be a heading of an appropriate level.'
      );
    }
    if (level !== this.level) {
      this.level = level;
    }
    this.headingElement.innerHTML = lightHeading.innerHTML;
    lightHeading.parentNode.removeChild(lightHeading);
  }

  handleClick = () => {
    this.expanded = !this.expanded;
    this.scaleExpand.emit({ expanded: this.expanded });
  };

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}

        <div
          class={this.getCssClassMap()}
          part={classNames('base', this.expanded && 'expanded')}
        >
          <h2
            aria-level={this.level}
            class="collapsible__heading"
            part="heading"
          >
            <button
              id={this.headingId}
              class="collapsible__button"
              part="button"
              onClick={this.handleClick}
              aria-expanded={this.expanded ? 'true' : 'false'}
              aria-controls={this.panelId}
            >
              <scale-icon-navigation-collapse-down
                size={16}
                decorative
                class="collapsible__icon"
                part={classNames('icon', this.expanded && 'expanded')}
              />
              <span
                ref={(el) => (this.headingElement = el)}
                part="button-text"
              />
            </button>
          </h2>
          <div
            id={this.panelId}
            role="region"
            aria-labelledby={this.headingId}
            hidden={!this.expanded}
            class="collapsible__content"
            part="content"
          >
            <slot />
          </div>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames('collapsible', this.expanded && 'collapsible--expanded');
  }
}
