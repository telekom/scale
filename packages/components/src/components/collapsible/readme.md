# scale-collapsible



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                    | Type                | Default     |
| -------------- | --------------- | ------------------------------ | ------------------- | ----------- |
| `expanded`     | `expanded`      | Set to `true` to expand        | `boolean`           | `undefined` |
| `headingLevel` | `heading-level` | Default aria-level for heading | `number`            | `2`         |
| `iconLocation` | `icon-location` |                                | `"left" \| "right"` | `'left'`    |
| `styles`       | `styles`        | (optional) Injected CSS styles | `string`            | `undefined` |


## Events

| Event          | Description                                                                                        | Type                                  |
| -------------- | -------------------------------------------------------------------------------------------------- | ------------------------------------- |
| `scale-expand` | Emitted so parent <scale-accordion> knows about it                                                 | `CustomEvent<CollapsibleEventDetail>` |
| `scaleExpand`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<CollapsibleEventDetail>` |


## Shadow Parts

| Part            | Description |
| --------------- | ----------- |
| `"button"`      |             |
| `"button-text"` |             |
| `"content"`     |             |
| `"heading"`     |             |


## Dependencies

### Depends on

- [scale-icon-navigation-collapse-down](../icons/navigation-collapse-down)

### Graph
```mermaid
graph TD;
  scale-collapsible --> scale-icon-navigation-collapse-down
  style scale-collapsible fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
