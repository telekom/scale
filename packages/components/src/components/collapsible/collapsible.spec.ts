import { newSpecPage } from '@stencil/core/testing';
import { Collapsible } from './collapsible';

describe('TextField', () => {
  let page;
  beforeEach(async () => {
    page = await newSpecPage({
      components: [Collapsible],
      html: `<scale-collapsible></scale-collapsible>`,
    });
  });
  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });
  it('should emit on click', async () => {
    const clickSpy = jest.fn();
    page.doc.addEventListener('scaleExpand', clickSpy);
    const buttonElement = page.root.shadowRoot.querySelector('button');
    buttonElement.click();
    await page.waitForChanges();
    expect(clickSpy).toHaveBeenCalled();
  });
});
