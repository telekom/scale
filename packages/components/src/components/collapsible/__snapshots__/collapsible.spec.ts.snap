// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should match snapshot 1`] = `
<scale-collapsible>
  <mock:shadow-root>
    <div class="collapsible" part="base">
      <h2 aria-level="2" class="collapsible__heading" part="heading">
        <button aria-controls="collapsable-panel-0" aria-expanded="false" class="collapsible__button" id="collapsable-heading-0" part="button">
          <scale-icon-navigation-collapse-down class="collapsible__icon" decorative="" part="icon" size="16"></scale-icon-navigation-collapse-down>
          <span part="button-text"></span>
        </button>
      </h2>
      <div aria-labelledby="collapsable-heading-0" class="collapsible__content" hidden="" id="collapsable-panel-0" part="content" role="region">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
</scale-collapsible>
`;
